executor/adapter.go

```diff
@@ -561,18 +565,121 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 	}, nil
 }
 
-func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor) error {
+func (a *ExecStmt) handleStmtForeignKeyTrigger(ctx context.Context, fkCtx FKCascadeContext, e Executor, isPessimistic bool) error {
+	if fkCtx.hasCascades {
+		// If the ExecStmt has foreign key cascade to be executed, we need call `StmtCommit` to commit the ExecStmt itself
+		// change first.
+		// Since `UnionScanExec` use `SnapshotIter` and `SnapshotGetter` to read txn mem-buffer, if we don't  do `StmtCommit`,
+		// then the fk cascade executor can't read the mem-buffer changed by the ExecStmt.
+		a.Ctx.StmtCommit()
+	}
+	err := a.handleForeignKeyTrigger(ctx, e, isPessimistic, 1)
+	if err != nil {
+		err1 := a.handleFKTriggerError(fkCtx)
+		if err1 != nil {
+			return errors.Errorf("handle foreign key trigger error failed, err: %v, original_err: %v", err1, err)
+		}
+		return err
+	}
+	if fkCtx.savepointName != "" {
+		a.Ctx.GetSessionVars().TxnCtx.ReleaseSavepoint(fkCtx.savepointName)
+	}
+	return nil
+}
+
+var maxForeignKeyCascadeDepth = 15
+
+func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, isPessimistic bool, depth int) error {
 	exec, ok := e.(WithForeignKeyTrigger)
 	if !ok {
 		return nil
 	}
+	originStmtCtx := a.Ctx.GetSessionVars().StmtCtx
+	defer func() {
+		// Reset to the original stmtCtx.
+		a.Ctx.GetSessionVars().StmtCtx = originStmtCtx
+	}()
 	fkChecks := exec.GetFKChecks()
 	for _, fkCheck := range fkChecks {
 		err := fkCheck.doCheck(ctx)
 		if err != nil {
 			return err
 		}
 	}
+	fkCascades := exec.GetFKCascades()
+	for _, fkCascade := range fkCascades {
+		err := a.handleForeignKeyCascade(ctx, fkCascade, isPessimistic, depth)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *ExecStmt) handleForeignKeyCascade(ctx context.Context, fkc *FKCascadeExec, isPessimistic bool, depth int) error {
+	if len(fkc.fkValues) == 0 {
+		return nil
+	}
+	if depth > maxForeignKeyCascadeDepth {
+		return ErrForeignKeyCascadeDepthExceeded.GenWithStackByArgs(maxForeignKeyCascadeDepth)
+	}
+	e, err := fkc.buildExecutor(ctx)
+	if err != nil || e == nil {
+		return err
+	}
+	if err := e.Open(ctx); err != nil {
+		terror.Call(e.Close)
+		return err
+	}
+	_, _, e, err = a.handleNoDelay(ctx, e, isPessimistic)
+	if err != nil {
+		return err
+	}
```

Comment:
I'm not quite clear about the previous state assumptions, maybe it's better to add some descriptions to explain the safety, which states of a statement execution may change or may not change. 
For example, should the executions refresh the `for_update_ts` or they should always use the same one?

Reply:
Great, maybe I should reuse `handleNoDelay` here, write an `handleFKExecutor` method to handle it maybe better.

Reply:
I have refactored the logic here, PTAL